whats-that-typeclass-functor/article.md

```diff
@@ -0,0 +1,467 @@
+You might already know the `map` function for lists:
+
+```haskell
+> map (+1) [1, 2, 3]
+[2, 3, 4]
+```
+
+It takes a function of type `a -> b` and applies it to each element of a list of type `[a]`. The values change, but the context (`[]`) remains the same.
+
+Transforming a value inside a fixed context like that is common in Haskell.
+
+For example, the optional data type ([`Maybe`](https://serokell.io/blog/algebraic-data-types-in-haskell#maybe)) could also be ‘mapped’ by trying to apply a function to the wrapped value without changing the context.
+
+In Haskell, we have a typeclass providing common functionality for such an action. It’s called Functor.
+
+After reading this article, you will:
+
+- know what Functor is in Haskell;
+- understand how to define and use your own Functor instances;
+- see why functors are useful and where they could be used;
+- be able to consolidate your knowledge with exercises.
+
+## How to generalize `map`
+
+Look at the type signature of `map`:
+
+```haskell
+map :: (a -> b) -> [a] -> [b]
+```
+
+It takes a function `a -> b`. Then, it uses that function to turn its second argument (of type `[a]`) into a list of `b` — `[b]`.
+
+Now imagine we’re using the same type of function — `a -> b` — to change the contents of  `Maybe`. For now, it doesn’t matter how exactly we’d do it. What type would such a `map'` function have?
+
+<details>
+<summary>The type of <code>map'</code> modifying <code>Maybe</code></summary>
+
+```haskell
+map' :: (a -> b) -> Maybe a -> Maybe b
+```
+<hr>
+</details>
+
+Finally, we can write a `map'` for an arbitrary data type `f` with one type argument `a` — `f a`. By analogy, can you figure out how it would look like?
```

Comment:
It might be better if we admit here upfront that this `map` function cannot work without knowing the innards of the type it is acting upon (and hence the reliance on a Typeclass), Or the reader might get that "magical Haskell" feeling here, making them lose all hope of ever groking the concept of the Functor

Reply:
True. 

I guess the phrase "we can write a map for arbitrary data type" is in this sense, wrong?

Reply:
We can generalize this type signature for an arbitrary data type `f` with one ... 


^ Perhaps this would be better for time being?

Reply:
>I guess the phrase "we can write a map for arbitrary data type" is in this sense, wrong?

I think it could be made more accurate. We cannot write a `map` function for any type, but just those that has defined how to `fmap` over themselves (via `Functor` Typeclass). So `Functor` instance is the generalization, and `map` function can work for many types because it uses this generalized interface.

Reply:
>we can also write map for type class Extract which has a method

Yea, I meant there should be some code that provides information or an implements an interface that can be used to do the mapping.

Reply:
Rephrased to avoid inaccurate wording. In addition, the note like following could be added here:

> Finally, we can generalize type signature of the `map'` for some* data type `f` with one type argument `a` — `f a`.

...

\* Not every data type could be a functor ...

However, I'm not sure if it's worth doing 🤔

Reply:
Actually, at this point it would be nice to mention that you cannot really implement a polymorphic fmap, you need a typeclass to guard and provide type-specific implementations. 🤔

Reply:
Rewrote